File: web/js/bellmanford.js
"use strict";

window.BellmanFord = {
    INF: Math.floor(Number.MAX_SAFE_INTEGER / 2),

    solve(costMatrix, capacityMatrix, source) {
        const n = costMatrix.length;
        const dist = new Array(n).fill(this.INF);
        const parent = new Array(n).fill(-1);
        dist[source] = 0;
        let iterations = 0;
        let hasNegativeCycle = false;

        // Relax edges n-1 times
        for (let iter = 0; iter < n - 1; iter++) {
            let updated = false;
            for (let u = 0; u < n; u++) {
                if (dist[u] >= this.INF) continue;
                for (let v = 0; v < n; v++) {
                    if (capacityMatrix[u][v] > 0) {
                        const newDist = dist[u] + costMatrix[u][v];
                        if (newDist < dist[v]) {
                            dist[v] = newDist;
                            parent[v] = u;
                            updated = true;
                        }
                    }
                }
            }
            iterations++;
            if (!updated) break;
        }

        // Check for negative cycles
        for (let u = 0; u < n && !hasNegativeCycle; u++) {
            if (dist[u] >= this.INF) continue;
            for (let v = 0; v < n; v++) {
                if (capacityMatrix[u][v] > 0 &&
                    dist[u] + costMatrix[u][v] < dist[v]) {
                    hasNegativeCycle = true;
                    break;
                }
            }
        }

        return { dist, parent, hasNegativeCycle, iterations };
    },

    reconstructPath(parent, source, target) {
        if (parent[target] === -1 && target !== source) return [];
        const path = [];
        for (let v = target; v !== -1; v = parent[v])
            path.push(v);
        path.reverse();
        return path;
    }
};

File: bellmanford.cpp
#include "bellmanford.h"
#include <climits>
#include <algorithm>

static constexpr int INF = INT_MAX / 2;

BellmanFordResult BellmanFord::solve(
    const std::vector<std::vector<int>>& costMatrix,
    const std::vector<std::vector<int>>& capacityMatrix,
    int source)
{
    const int n = static_cast<int>(costMatrix.size());
    BellmanFordResult res;
    res.dist.assign(n, INF);
    res.parent.assign(n, -1);
    res.dist[source] = 0;

    // Relax edges n-1 times
    for (int iter = 0; iter < n - 1; ++iter) {
        bool updated = false;
        for (int u = 0; u < n; ++u) {
            if (res.dist[u] >= INF) continue;
            for (int v = 0; v < n; ++v) {
                if (capacityMatrix[u][v] > 0) {
                    int newDist = res.dist[u] + costMatrix[u][v];
                    if (newDist < res.dist[v]) {
                        res.dist[v] = newDist;
                        res.parent[v] = u;
                        updated = true;
                    }
                }
            }
        }
        ++res.iterations;
        if (!updated) break;
    }

    // Check for negative cycles
    for (int u = 0; u < n && !res.hasNegativeCycle; ++u) {
        if (res.dist[u] >= INF) continue;
        for (int v = 0; v < n; ++v) {
            if (capacityMatrix[u][v] > 0 &&
                res.dist[u] + costMatrix[u][v] < res.dist[v]) {
                res.hasNegativeCycle = true;
                break;
            }
        }
    }

    return res;
}

std::vector<int> BellmanFord::reconstructPath(const std::vector<int>& parent,
                                              int source, int target)
{
    std::vector<int> path;
    if (parent[target] == -1 && target != source)
        return path;

    for (int v = target; v != -1; v = parent[v])
        path.push_back(v);

    std::reverse(path.begin(), path.end());
    return path;
}
